### ISSUE TITLE
Crash when building Go package 'runtime'

### ISSUE THREAD

**Issue by melonedo** [plctlab/gollvm#9]
On ubuntu-20.04.2-preinstalled-server-riscv64 (qemu):
```shell
cmake -DCMAKE_BUILD_TYPE=Debug -DGOLLVM_USE_SPLIT_STACK=OFF -DGOLLVM_DEFAULT_LINKER=bfd -DLLVM_USE_LINKER=bfd -DLLVM_TARGETS_TO_BUILD="RISCV" -G Ninja ../llvm-project/llvm -D CMAKE_INSTALL_PREFIX=../install-debug/
ninja gollvm
```
Error:
```plain
[1/1215] Building Go package 'runtime' (PIC)                                    
FAILED: tools/gollvm/libgo/.pic/runtime.o                                       
cd /home/ubuntu/build-debug/tools/gollvm/libgo && /usr/bin/cmake -E make_directory ././.pic && /home/ubuntu/build-debug/./bin/llvm-goc -c -o /home/ubuntu/build-debug/tools/gollvm/libgo/./.pic/runtime.o -fPIC -fgo-pkgpath=runtime -fgo-c-header=runtime.inc.raw -fgo-compiling-runtime -fno-split-stack -I . /home/ubuntu/llvm-project/llvm/tools/gollvm/gofrontend/libgo/go/runtime/alg.go 
[list of source files] /home/ubuntu/build-debug/tools/gollvm/libgo/goroot.go 
PromoteIntegerOperand Op #0: t23: i64 = EH_DWARF_CFA Constant:i32<0>, cgo_gccgo.go:86:16

Do not know how to promote this operator's operand!
UNREACHABLE executed at /home/ubuntu/llvm-project/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:1567!                                                  
#0 0x0000002abf47bf68 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (.localalias) /home/ubuntu/llvm-project/llvm/lib/Support/Unix/Signals.inc:565:22
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.                                                     
Stack dump:
0.      Program arguments: /home/ubuntu/build-debug/./bin/llvm-goc
[see above]
/home/ubuntu/build-debug/tools/gollvm/libgo/goroot.go                                                                     
1.      Running pass 'Function Pass Manager' on module 'gomodule'.              
2.      Running pass 'RISCV DAG->DAG Pattern Instruction Selection' on function 
'@syscall.CgocallBack'
Aborted (core dumped)
```

**Comment by luxufan** [comment]
It was caused by your llvm source code couldn't custom lowering EH_DWARF_CFA node for riscv target.

https://github.com/plctlab/llvm-project/tree/llvm-for-gollvm This branch has supported custom lowering EH_DWARF_CFA node

**Comment by melonedo** [comment]
Thanks! That worked. But I have a new failure now...
```
[1938/1948] Building go program vet
FAILED: tools/gollvm/gotools/vet 
cd /home/ubuntu/build-debug/tools/gollvm/gotools && /home/ubuntu/build-debug/./bin/llvm-goc -o /home/ubuntu/build-debug/tools/gollvm/gotools/vet vet_.o -fno-split-stack -I /home/ubuntu/build-debug/tools/gollvm/libgo -L /home/ubuntu/build-debug/tools/gollvm/libgo /home/ubuntu/build-debug/tools/gollvm/libgo/libgotool.a
/usr/bin/ld.bfd: /home/ubuntu/build-debug/tools/gollvm/libgo/libgo.so: undefined reference to `__atomic_fetch_or_1'
/usr/bin/ld.bfd: /home/ubuntu/build-debug/tools/gollvm/libgo/libgo.so: undefined reference to `__atomic_fetch_and_1'
```
Do I need to specify `-latomic` as in [this CMake issue](https://gitlab.kitware.com/cmake/cmake/-/issues/20895)?

**Comment by melonedo** [comment]
This is resolved with a more recent toolchain on Ubuntu 22.04 instead of 20.04
